From b9e94a9d620a0f5844f5ab75651b55445d12243a Mon Sep 17 00:00:00 2001
From: Sergey Shelukhin <sershe@apache.org>
Date: Tue, 12 Jul 2016 21:31:43 -0700
Subject: [PATCH 0675/1164] CDH-36924 HIVE-14195 : HiveMetaStoreClient
 getFunction() does not throw
 NoSuchObjectException (Peter Vary, reviewed by
 Sergey Shelukhin)

Change-Id: I59b566f0e204a16e21b6d2be51d56832faed5a5c
---
 .../hadoop/hive/metastore/TestHiveMetaStore.java   |    2 +-
 .../hadoop/hive/metastore/HiveMetaStore.java       |    3 +++
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/itests/hive-unit/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java b/itests/hive-unit/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java
index 423623a..f242e97 100644
--- a/itests/hive-unit/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java
+++ b/itests/hive-unit/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java
@@ -2627,7 +2627,7 @@ public void testSimpleFunction() throws Exception {
       boolean gotException = false;
       try {
         func = client.getFunction(dbName, "nonexistent_func");
-      } catch (MetaException e) {
+      } catch (NoSuchObjectException e) {
         // expected failure
         gotException = true;
       }
diff --git a/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java b/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java
index 7145ee5..8b002d2 100644
--- a/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java
+++ b/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java
@@ -5760,6 +5760,9 @@ public Function get_function(String dbName, String funcName)
           throw new NoSuchObjectException(
               "Function " + dbName + "." + funcName + " does not exist");
         }
+      } catch (NoSuchObjectException e) {
+        ex = e;
+        rethrowException(e);
       } catch (Exception e) {
         ex = e;
         throw newMetaException(e);
-- 
1.7.9.5

